### PR TITLE
Feature/sc 377004/add raster loader support for snowflake

### PR DESCRIPTION
**Pull request by volaya**
## Proposed Changes

This PR adds support for Snowflake

A refactoring of the main scripts has also been made, to leave a cleaner and more extensible code, which allows to reuse code between platforms.

### CODE REVIEW

**Comment by vdelacruzb** [comment]
I'm struggling still to achieve loading a raster into snowflake. When executing the upload command I receive a crash because of the snowflake python connector not being installed.
I've looked for related threads and found [this](https://github.com/snowflakedb/snowflake-connector-python/issues/1331) it seems that we shouldn't have any file called `snowflake.py`. I've tried locally to replace file or class name called snowflake by other names and I don't get that error.

Also got a:
```
Error: Error uploading to Snowflake: 090105 (22000): Cannot perform CREATE TEMPSTAGE. This session does not have a current database. Call 'USE DATABASE', or use a qualified name.
```
So we might have to set the database as default when uploading the table.

**Comment by vdelacruzb** [comment]
I've just noticed that in the ticket the parameter --role was included:

```
carto snowflake upload \
  --file_path /path/to/my/raster/file.tif \
  --account my-account \
  --database my-database \
  --schema my-schema \
  --table my-table \
  --role my-role \
  --user my-user \
  --password $MY_PASSWORD \
  --overwrite
```

It would be nice to include it as it will define who is the owner of the table in snowflake once created.

**Comment by vdelacruzb** [comment]
I've done some changes since I've discovered that the chunks are getting constantly overwritten. That's why I was getting too small table results during the QA.
- BQ: I like the approach of using `writeDisposition` but it seems that the jobs are not executed in the proper order so it first appends records and at the end truncate. So for the moment I got back to the previous behaviour removing the input table when `--overwrite`.
-SF: I updated the script to only overwrite records in the first iteration.

**Review by vdelacruzb — APPROVED** [review]
Nice LGTM.

Just notice I've done some updates in the code to not overwrite chunks.

Also as I was having when appending multiple rasters with nodata=NaN. I've explicitely updated this value to null, as I saw in a previously updated table `cartodb-data-engineering-team.jarroyo.usa_pop_raster`.

File: docs/source/user_guide/cli.rst [hunk 1]
```diff
@@ -7,25 +7,45 @@ Most functions of the Raster Loader are accessible through the carto
 command-line interface (CLI). To start the CLI, use the ``carto`` command in a
 terminal.
 
-Currently, Raster Loader allows you to upload a local raster file to a BigQuery table.
-You can also download and inspect a raster file from a BigQuery table.
+Currently, Raster Loader allows you to upload a local raster file to a BigQuery or Snowflake table.
+You can also download and inspect a raster file from a BigQuery or Snowflake table.
```
Comment (Jesus89) [inline docs/source/user_guide/cli.rst:11]: Describe does not exist anymore, so we can remove this line

File: docs/source/user_guide/cli.rst [hunk 2]
```diff
@@ -139,11 +179,11 @@ of 2000 rows:
 
 
 
-Inspecting a raster file on BigQuery
+Inspecting a raster file
```
Comment (Jesus89) [inline docs/source/user_guide/cli.rst:182]: Describe does not exist anymore, so we can remove this section

File: raster_loader/io/bigquery.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    from google.cloud import bigquery
+except ImportError:  # pragma: no cover
+    _has_bigquery = False
+else:
+    _has_bigquery = True
+
+from raster_loader.io.datawarehouse import DataWarehouse
+
+
+class BigQuery(DataWarehouse):
```
Comment (Jesus89) [inline raster_loader/io/bigquery.py:29]: I would rename it as `BigQueryConnection`

File: raster_loader/io/datawarehouse.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
+import pandas as pd
+
+from typing import List
+
+
+class DataWarehouse:
```
Comment (Jesus89) [inline raster_loader/io/datawarehouse.py:6]: I would rename it as `DataWarehouseConnection`

File: raster_loader/io/snowflake.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+try:
+    from snowflake.connector.pandas_tools import write_pandas
+    import snowflake.connector
+except ImportError:  # pragma: no cover
+    _has_snowflake = False
+else:
+    _has_snowflake = True
+
+
+class Snowflake(DataWarehouse):
```
Comment (Jesus89) [inline raster_loader/io/snowflake.py:32]: I would rename it as `SnowflakeConnection`

Comment (vdelacruzb) [reply raster_loader/io/snowflake.py]: Updated

File: raster_loader/io/datawarehouse.py (hunk 4 above)
Comment (vdelacruzb) [reply raster_loader/io/datawarehouse.py]: Updated

File: raster_loader/io/bigquery.py (hunk 3 above)
Comment (vdelacruzb) [reply raster_loader/io/bigquery.py]: Updated

**Review by Jesus89 — APPROVED** [review]
LGTM

**Comment by Jesus89** [comment]
Let's release `0.5.0` after merging this PR

**Pull request merged**